### ISSUE TITLE
Invalid Signature on Debian Binaries for 2.0.1

### ISSUE THREAD

**Issue by DiagonalArg** [zcash/zcash#3634]
Ubuntu 16.04
Zcash is version 2.0.1

On my most recent update, I started getting this error:

```
The following signatures were invalid: KEYEXPIRED 1539886450
```

So I followed the [Update Signing Keys](https://zcash.readthedocs.io/en/latest/rtd_pages/install_debian_bin_packages.html#updating-signing-keys) guide.  At first I was doing it using:

```
sudo apt-key del 1539886450
```

since that was the key that `apt-get update` reported as expired; but that had no effect on the error.  So I followed the guide verbatim, using:

```
sudo apt-key del 63C4A2169C1B2FA2
```

Then I got a new error:

```
The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2
```

That error appears similar to [#2590](https://github.com/zcash/zcash/issues/2590).  Note that I did indeed obtain the key via `wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -`, and even checked what I was getting with `curl -O`

```
$ sudo apt-key list

[... snip]

pub 4096R/66E279CD 2018-10-19 [expires: 2020-10-20]
uid Zcash Master Signing Key (Zcash Electric Coin Company)
sub 4096R/68AE9744 2018-10-19 [expires: 2020-10-20]

[... snip]
```

```
$ sudo apt-get update

[... snip]

Get:22 https://apt.z.cash jessie Release [2,288 B]                                                              
Get:23 http://us.archive.ubuntu.com/ubuntu xenial-backports InRelease [107 kB]                                  
Err:24 https://apt.z.cash jessie Release.gpg                                         
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2

[... snip]

W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://apt.z.cash jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2

[... snip]
```

```
$ sudo apt-get -o Debug::Acquire::gpgv=true update

Ign:20 https://apt.z.cash jessie InRelease 

[... snip]

Get:23 https://apt.z.cash jessie Release [2,288 B]
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /var/lib/apt/lists/apt.z.cash_dists_jessie_Release.gpg /var/lib/apt/lists/apt.z.cash_dists_jessie_Release
Read: [GNUPG:] ERRSIG 63C4A2169C1B2FA2 1 8 00 1539619038 9

Got ERRSIG
Read: [GNUPG:] NO_PUBKEY 63C4A2169C1B2FA2

Got NO_PUBKEY
gpgv exited with status 2
Summary:
  Good:
  Bad:
  Worthless:
  SoonWorthless:
  NoPubKey: NO_PUBKEY 63C4A2169C1B2FA2
,
Err:24 https://apt.z.cash jessie Release.gpg
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /tmp/apt.sig.fCkHy7 /tmp/apt.data.zT1gDy

[... snip]

W: Failed to fetch https://apt.z.cash/dists/jessie/Release.gpg  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2

[... snip]
```

So I got the key back:

```
$ sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 63C4A2169C1B2FA2
Executing: /tmp/tmp.vHh8OopwV8/gpg.1.sh --keyserver
keyserver.ubuntu.com
--recv-keys
63C4A2169C1B2FA2
gpg: requesting key 9C1B2FA2 from hkp server keyserver.ubuntu.com
gpg: key 9C1B2FA2: public key "Zcash Master Signing Key" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
```

And am back to the original error:

```
$ sudo apt-get update
Err:24 https://apt.z.cash jessie Release.gpg                                                                     
  The following signatures were invalid: KEYEXPIRED 1539886450
```

```
$ sudo apt-get -o Debug::Acquire::gpgv=true update

[... snip]

Ign:16 https://apt.z.cash jessie InRelease
Read: [GNUPG:] SIG_ID fzTdB4Uv5EiuJd0/MrXqmQseTjk 2018-10-19 1539967988

Read: [GNUPG:] GOODSIG EF8B479E2DC1389C Jitsi <dev@jitsi.org>

Got GOODSIG, key ID: EF8B479E2DC1389C
Read: [GNUPG:] VALIDSIG 66A9CD0595D6AFA247290D3BEF8B479E2DC1389C 2018-10-19 1539967988 0 4 0 1 10 01 66A9CD0595D6AFA247290D3BEF8B479E2DC1389C

Got trusted VALIDSIG, key ID: 66A9CD0595D6AFA247290D3BEF8B479E2DC1389C
gpgv exited with status 0
Summary:
  Good: GOODSIG EF8B479E2DC1389C,
  Bad:
  Worthless:
  SoonWorthless:
  NoPubKey:
apt-key succeeded
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /tmp/apt.sig.AWjUNy /tmp/apt.data.XpuhZP
Read: [GNUPG:] SIG_ID y5200frSIRv0qPnZ8csCMwx1Kqw 2018-10-16 1539716054

Read: [GNUPG:] GOODSIG D46F45428842CE5E Launchpad PPA for Bitcoin

Got GOODSIG, key ID: D46F45428842CE5E
Read: [GNUPG:] VALIDSIG C70EF1F0305A1ADB9986DBD8D46F45428842CE5E 2018-10-16 1539716054 0 4 0 1 10 01 C70EF1F0305A1ADB9986DBD8D46F45428842CE5E

Got trusted VALIDSIG, key ID: C70EF1F0305A1ADB9986DBD8D46F45428842CE5E
gpgv exited with status 0
Summary:
  Good: GOODSIG D46F45428842CE5E,
  Bad:
  Worthless:
  SoonWorthless:
  NoPubKey:
apt-key succeeded

[... snip]

Get:18 https://apt.z.cash jessie Release [2,288 B]
Read: [GNUPG:] SIG_ID 0o4zMC0maVvTy9ZRCLTpfg+G4hc 2018-08-30 1535669272

Read: [GNUPG:] GOODSIG BE99746B2FB21B35 Launchpad PPA for bladeRF

Got GOODSIG, key ID: BE99746B2FB21B35
Read: [GNUPG:] VALIDSIG 188FE585DD24922CE9CD1EE9BE99746B2FB21B35 2018-08-30 1535669272 0 4 0 1 10 01 188FE585DD24922CE9CD1EE9BE99746B2FB21B35

Got trusted VALIDSIG, key ID: 188FE585DD24922CE9CD1EE9BE99746B2FB21B35
gpgv exited with status 0
Summary:
  Good: GOODSIG BE99746B2FB21B35,
  Bad:
  Worthless:
  SoonWorthless:
  NoPubKey:
apt-key succeeded
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /var/lib/apt/lists/apt.z.cash_dists_jessie_Release.gpg /var/lib/apt/lists/apt.z.cash_dists_jessie_Release
Read: [GNUPG:] KEYEXPIRED 1539886450

Got KEYEXPIRED!
Read: [GNUPG:] SIGEXPIRED deprecated-use-keyexpired-instead

Read: [GNUPG:] SIG_ID ojNEQ8+zSQmFTjuCaILYK8oyMLc 2018-10-15 1539619038

Read: [GNUPG:] EXPKEYSIG 63C4A2169C1B2FA2 Zcash Master Signing Key

Read: [GNUPG:] VALIDSIG F1E214037E94E950BA8577B263C4A2169C1B2FA2 2018-10-15 1539619038 0 4 0 1 8 00 F1E214037E94E950BA8577B263C4A2169C1B2FA2

Got trusted VALIDSIG, key ID: F1E214037E94E950BA8577B263C4A2169C1B2FA2
gpgv exited with status 0
Summary:
  Good:
  Bad:
  Worthless: KEYEXPIRED 1539886450
,
  SoonWorthless:
  NoPubKey:

[... snip]

Err:19 https://apt.z.cash jessie Release.gpg
  The following signatures were invalid: KEYEXPIRED 1539886450
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /tmp/apt.sig.rjhsYl /tmp/apt.data.muJUtF

[... snip]

W: Failed to fetch https://apt.z.cash/dists/jessie/Release.gpg  The following signatures were invalid: KEYEXPIRED 1539886450

[... snip]
```

**Comment by DiagonalArg** [comment]
@ianamunoz - No, you have misunderstood.  

I followed the instructions in the troubleshooting guide, removing the key about which apt-get was complaining (KEYEXPIRED 1539886450), which had *no effect*.  I then used the troubleshooting instructions verbatim and removed the key referred to there (63C4A2169C1B2FA2), which had the effect of producing a different error (NO_PUBKEY 63C4A2169C1B2FA2).  I then repulled that second key, since it was clearly complaining about not having it, and found myself back at the first error.  Each time I did the cleaning as described in the troubleshooting guide.

**Comment by garethtdavies** [comment]
You definitely removed the local source list as per the below?

```
sudo rm /etc/apt/sources.list.d/zcash.list
sudo apt-get update
wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
echo "deb [arch=amd64] https://apt.z.cash/ jessie main" | sudo tee /etc/apt/sources.list.d/zcash.list
sudo apt-get update
```

**Comment by DiagonalArg** [comment]
@garethtdavies - yes, repeatedly.  Exactly as in [the guide](https://zcash.readthedocs.io/en/latest/rtd_pages/install_debian_bin_packages.html#updating-signing-keys) I linked above.

**Comment by garethtdavies** [comment]
and what does your `apt key-list` look like after running:

```
sudo apt-key del AEFD26F966E279CD
wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
```

**Comment by ianamunoz** [comment]
@DiagonalArg I only mention to remove the list items and do a sucessful `apt-get update` because in the middle of your logs you do a: `sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 63C4A2169C1B2FA2` which would cause you to pull down the previously expired key again.

You need to make sure you are pulling the new key from `apt.z.cash`.

**Comment by ianamunoz** [comment]
```
Err:19 https://apt.z.cash jessie Release.gpg
  The following signatures were invalid: KEYEXPIRED 1539886450
inside VerifyGetSigners
Preparing to exec:  /usr/bin/apt-key --quiet --readonly verify --status-fd 3 /tmp/apt.sig.rjhsYl /tmp/apt.data.muJUtF
```

at the end of your logs shows you still have the old key. Remove the old key. Remove the old apt meta data release information. Then you should be able to re-add the new key and successfully update.

**Comment by garethtdavies** [comment]
@ianamunoz Surely the old key isn't the issue as the release is signed with the new one so having the old wouldn't cause any issues as it'd never be used and simply good hygiene to remove it? Just removing all keys and running sudo apt-get update should prompt you for the key `NO_PUBKEY AEFD26F966E279CD` so as you say above it's the release information is out of date and the published fix should sort this.

**Comment by DiagonalArg** [comment]
Just so everyone is clear (@ianamunoz), I ran:

```
sudo apt-key del 1539886450
sudo rm /etc/apt/sources.list.d/zcash.list
sudo apt-get update
wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
sudo apt-get clean
echo "deb [arch=amd64] https://apt.z.cash/ jessie main" | sudo tee /etc/apt/sources.list.d/zcash.list
sudo apt-get update
```
The error both before and after running this is:

```
The following signatures were invalid: KEYEXPIRED 1539886450
```

That error does not appear after the intermediate `apt-get update`.  Note that as @garethtdavies suggests, this time I included the `apt-get clean`.

`apt-key list` after running this, contains:

```
pub   4096R/66E279CD 2018-10-19 [expires: 2020-10-20]
uid                  Zcash Master Signing Key (Zcash Electric Coin Company)
sub   4096R/68AE9744 2018-10-19 [expires: 2020-10-20]
```

**Comment by DiagonalArg** [comment]
@garethtdavies - 

```
$ sudo apt-key del AEFD26F966E279CD
$ sudo apt-key list

pub   4096R/9C1B2FA2 2016-10-18 [expired: 2018-10-18]
uid                  Zcash Master Signing Key

$ wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
$ sudo apt-key list

pub   4096R/9C1B2FA2 2016-10-18 [expired: 2018-10-18]
uid                  Zcash Master Signing Key

pub   4096R/66E279CD 2018-10-19 [expires: 2020-10-20]
uid                  Zcash Master Signing Key (Zcash Electric Coin Company)
sub   4096R/68AE9744 2018-10-19 [expires: 2020-10-20]
```

**Comment by garethtdavies** [comment]
@DiagonalArg Those keys look fine. Let's isolate the issue. Just remove them both:

`sudo apt-key delete 9C1B2FA2`
`sudo apt-key delete 66E279CD`

Now when you run `sudo apt-get update` what key does it say is missing. If it's the former (9C1B2FA2) then this is certainly an issue of it not getting the latest release information (though I have no idea why).

**Comment by DiagonalArg** [comment]
@garethdavies - ok,

```
sudo apt-key del 9C1B2FA2
sudo apt-key del 66E279CD
mv /etc/apt/sources.list.d/zcash.list ~
sudo apt-get update
mv ~/zcash.list /etc/apt/sources.list.d/
sudo apt-get update
```

The missing key is now:

`The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 63C4A2169C1B2FA2`

`sudo apt-key` list returns no key associated with zcash.

```
wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
mv /etc/apt/sources.list.d/zcash.list ~
sudo apt-get update
mv ~/zcash.list /etc/apt/sources.list.d/
sudo apt-get update
```

Same error is returned.  `sudo apt-key` list returns one key associated with zcash:

```
pub   4096R/66E279CD 2018-10-19 [expires: 2020-10-20]
uid                  Zcash Master Signing Key (Zcash Electric Coin Company)
sub   4096R/68AE9744 2018-10-19 [expires: 2020-10-20]
```

**Comment by ianamunoz** [comment]
@DiagonalArg  I made a comment in the community chat. Do a:
`sudo ls /etc/apt/sources.list.d/` 

and share that list of resources. The apt information for the old key is still persistent so I don't think we succeeded in removing it.

**Comment by DiagonalArg** [comment]
Better I had already grepped the whole /etc/apt directory and there's nothing other than that one file containing the word "cash".

**Comment by nils-ballmann** [comment]
After the deletion of the key and its removal of `zcash.list` there needs to also be a `sudo apt-get clean` in there before you re-add the key and the `zcash.list`. At least this fixed the same issue for me.

**Comment by DiagonalArg** [comment]
@nils-ballmann  - If you look above, you will see that I tried that some time back.

**Comment by nils-ballmann** [comment]
Hm, this was eight days ago. For me it worked yesterday. So maybe they changed something. Maybe retry?

apt said it didn't have the right old master key for `Release.gpg`, but when I downloaded the file directly and checked the signature it wanted the new master key. So there must be a cached version of `Release.gpg` which apt isn't redownloading...

**Comment by daira** [comment]
Honestly it's hopeless to try to continue using the same key once a revocation cert for it has been sent to the key servers. We'll have to rotate the key again, and we should do so soon, since the current workaround isn't working for all users.

**Comment by annonch** [comment]
I had this same problem. It seemed that even when trying to delete the key after getting

`The following signatures were invalid: KEYEXPIRED 1539886450` 

I assumed that was the key id to delete but when listing keys I still saw a key

`sudo apt-key delete 66E279CD`

then to fix following @DiagonalArg 's instructions. 

`sudo rm /etc/apt/sources.list.d/zcash.list
sudo apt-get update
wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -
sudo apt-get clean
echo "deb [arch=amd64] https://apt.z.cash/ jessie main" | sudo tee /etc/apt/sources.list.d/zcash.list
sudo apt-get update`

Only odd thing is I needed to download the zcash params and that seemed to have a 301 redirect https://pastebin.com/ejJSV1RG

**Comment by ianamunoz** [comment]
@annonch  that 301 is intentionally in place to redirect you to the cloudfront CDN to increase availability and speed to download. Are you unable to run the `zcash-fetch-params`? That would be a separate issue then this.  (looks like you have a 206 there so you have downloaded some of it?)

**Comment by annonch** [comment]
I have no further issue. I just wanted to add another point of data with regard to the key expired issue. thanks!

**Comment by nils-ballmann** [comment]
Maybe removing `/var/lib/apt/lists/*` clears the apt cache and fixes this.

**Comment by DiagonalArg** [comment]
@nils-ballmann - Ok, now *that* solved the problem.  Finally.  

Oddly, apt-cache doesn't have an option to clear/clean/delete the cache.

Thanks.

**Comment by ianamunoz** [comment]
@nils-ballmann  Thanks for the that! It was hard for me to replicate so I didn't think of manually removing the cache. I kind of that an `apt-get clean` would have caught the cached packages. @DiagonalArg glad this got resolved for you finally. I am going to close this issue.

**Comment by daira** [comment]
@nils-ballmann wrote:
> Maybe removing /var/lib/apt/lists/* clears the apt cache and fixes this.

This may work, but it's by no means a reasonable thing to ask users to do! Reopening so that we don't lose track of the problems this is still causing users.

**Comment by nils-ballmann** [comment]
> 
> 
> @nils-ballmann wrote:
> 
> > Maybe removing /var/lib/apt/lists/* clears the apt cache and fixes this.
> 
> This may work, but it's by no means a reasonable thing to ask users to do! Reopening so that we don't lose track of the problems this is still causing users.

While I'm with you, that it should be the last resort to do things like this, I don't know if zcash should be fixing `apt` issues...

**Comment by daira** [comment]
The right thing to do is not to fix the `apt` issues directly, but to roll the key again so that they don't occur.

**Comment by mms710** [comment]
This has been completed. Closing.

**Comment by DiagonalArg** [comment]
The problem is now back:

> W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://apt.z.cash jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY C2A798EF998940FA

**Comment by garethtdavies** [comment]
This is expected as the key was rotated so you need to add this new key. See the **Missing Public Key**  section here: https://zcash.readthedocs.io/en/latest/rtd_pages/install_debian_bin_packages.html but simply `wget -qO - https://apt.z.cash/zcash.asc | sudo apt-key add -` followed by another `apt-get update` will cure it.
